alpenhorn/update.py

```diff
@@ -36,104 +45,603 @@
+        if self.db.io_class != new.io_class:
+            return True
 
-    # Check the integrity of any questionable files (has_file=M)
-    update_node_integrity(node)
+        return False
+
+    def _parse_io_config(self, config_json: str | None) -> dict:
+        """Parse and return the I/O config.
+
+        Parameters
+        ----------
+        config_json : str or None
+            The I/O config JSON string
+
+        Returns
+        -------
+        io_config : dict
+            The parsed I/O config, or an empty dict() if `config_json`
+            was `None.  This value is also assigned to `self._io_config`.
+
+        Raises
+        ------
+        ValueError
+            `config_json` did not evaluate to a dict.
+        """
+        if config_json is None:
+            self._io_config = dict()
+        else:
+            self._io_config = json.loads(config_json)
 
-    # Delete any upwanted files to cleanup space
-    update_node_delete(node)
+            if not isinstance(self._io_config, dict):
+                raise ValueError(f'Invalid io_config: "{config_json}".')
 
-    # Process any regular transfers requests onto this node
-    update_node_requests(node)
+        return self._io_config
 
-    # TODO: bring back HPSS support
-    # Process any tranfers out of HPSS onto this node
-    # update_node_hpss_outbound(node)
+    def _get_io_class(self):
+        """Return the I/O class for our Storage object."""
+
+        # If no io_class is specified, the Default I/O classes are used
+        io_name = "Default" if self.db.io_class is None else self.db.io_class
+
+        # We assume StorageNode if not StorageGroup
+        if self.is_group:
+            obj_type = "StorageGroup"
+            io_suffix = "GroupIO"
+        else:
+            obj_type = "StorageNode"
+            io_suffix = "NodeIO"
+
+        # Separate package and module
+        dot = io_name.rfind(".")
+        if dot < 0:
+            # If there's no package, assume it's part of alpenhorn.io
+            io_module = "alpenhorn.io." + io_name.lower()
+            io_name += io_suffix
+        else:
+            # Otherwise, separate the module name for later
+            io_module = io_name[:dot]
+            io_name = io_name[dot + 1 :]
 
+        # Load the module
+        try:
+            module = importlib.import_module(io_module)
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                f'Unable to find I/O module "{io_module}" for {obj_type} {self.name}.'
+            ) from e
 
-def update_node_active(node):
-    """Check if a node is actually active in the filesystem"""
+        # Within the module, find the class
+        try:
+            class_ = getattr(module, io_name)
+        except AttributeError as e:
+            raise ImportError(
+                f'I/O class "{io_name}" not found in module "{io_module}". '
+                f"Required for {obj_type} {self.name}."
+            ) from e
```

Comment:
I would be pretty tempted to move this class resolution code into alpenhorn startup, i.e. build up a dict of enabled io class name -> implementation from the default alpenhorn ones and those provided by plugins.

I think having this in here pollutes the logic a bit, and it means that you could be potentially running imports within the main loop of alpenhorn which could cause issues. It feels like it would be much better to do this upfront when the service starts along with all the other config/plugin type stuff.

Reply:
I get what you're saying.

I think effectively, in the normal case where available nodes don't change over the course of an `alpenhornd` run, this already happens for nodes, because all node I/O instantiation happens at the top of the first update loop, before the node update loop starts, and then doesn't change after that.

This is different than the Info class modules, which would be loaded at a surprising time in the middle of an import task if they weren't preloaded on start-up (and so it makes a lot of sense to do the preload). So, I think it's less of an issue here for I/O classes.

I don't think that just blindly preloading all I/O classes is a good idea, because it would require installations to have all their weird third-party I/O modules installed, even on hosts where nodes using those weird I/O modules don't exist.

As a half-way solution, why don't I also move the group-I/O instantiation to the top of the loop (currently happens after node updates are done).  Then all the module loading is done early, at the start of the first update loop, except for cases where new nodes/groups show up later with hitherto unseen I/O classes.

Reply:
Yeah, your half-way solution seems reasonable, and this isn't the kind of structuring that is embedded so deeply it can't be changed later.

To clarify though, I wasn't expecting that you would pre-load or require all possible IO classes, just that you would specify them through the plugin system, and thus the local config would only need to specify (and thus load) IO classes for all the types of nodes it expects to see.